### PR TITLE
Add --broker-namespace flag to deploy-broker

### PR DESCRIPTION
**Pull request by donaldh**
Deploys broker to custom namespace. Submariner join uses broker namespace
from broker-info.subm

Signed-off-by: Donald Hunter <donald.hunter@redhat.com>

Implemented as a first step to supporting multiple "tenant" brokers in a single cluster and could be part of what is needed for #1040 – Support deployment within a custom namespace.

### CODE REVIEW

File: pkg/broker/ensure.go [hunk 1]
```diff
@@ -74,88 +74,88 @@ func Ensure(config *rest.Config, componentArr []string, crds bool) error {
...
+	_, err = CreateNewBrokerRoleBinding(clientset, submarinerBrokerClusterDefaultSA, submarinerBrokerClusterRole, namespace)
 	if err != nil && !apierrors.IsAlreadyExists(err) {
 		return fmt.Errorf("error creating the broker rolebinding: %s", err)
 	}
 	return nil
 }
 
 // CreateSAForCluster creates a new SA, and binds it to the submariner cluster role
-func CreateSAForCluster(clientset *kubernetes.Clientset, clusterID string) (*v1.Secret, error) {
+func CreateSAForCluster(clientset *kubernetes.Clientset, clusterID string, namespace string) (*v1.Secret, error) {
```
Comment (skitt) [inline pkg/broker/ensure.go]:
```suggestion
func CreateSAForCluster(clientset *kubernetes.Clientset, clusterID, namespace string) (*v1.Secret, error) {
```

File: pkg/broker/ensure.go [hunk 2]
```diff
@@ -182,25 +182,25 @@ func WaitForClientToken(clientset *kubernetes.Clientset, submarinerBrokerSA stri
...
 }
 
-func CreateOrUpdateBrokerAdminRole(clientset *kubernetes.Clientset) (created bool, err error) {
-	return utils.CreateOrUpdateRole(context.TODO(), clientset, SubmarinerBrokerNamespace, NewBrokerAdminRole())
+func CreateOrUpdateBrokerAdminRole(clientset *kubernetes.Clientset, namespace string) (created bool, err error) {
+	return utils.CreateOrUpdateRole(context.TODO(), clientset, namespace, NewBrokerAdminRole())
 }
 
-func CreateNewBrokerRoleBinding(clientset *kubernetes.Clientset, serviceAccount, role string) (brokerRoleBinding *rbac.RoleBinding,
+func CreateNewBrokerRoleBinding(clientset *kubernetes.Clientset, serviceAccount, role string, namespace string) (brokerRoleBinding *rbac.RoleBinding,
```
Comment (skitt) [inline pkg/broker/ensure.go]:
```suggestion
func CreateNewBrokerRoleBinding(clientset *kubernetes.Clientset, serviceAccount, role, namespace string) (brokerRoleBinding *rbac.RoleBinding,
```

File: pkg/broker/rbac.go [hunk 3]
```diff
@@ -110,7 +110,7 @@ func NewBrokerClusterRole() *rbacv1.Role {
 }
 
 // Create a role for to bind the cluster admin (subctl) SA
-func NewBrokerRoleBinding(serviceAccount, role string) *rbacv1.RoleBinding {
+func NewBrokerRoleBinding(serviceAccount, role string, namespace string) *rbacv1.RoleBinding {
```
Comment (skitt) [inline pkg/broker/rbac.go]:
```suggestion
func NewBrokerRoleBinding(serviceAccount, role, namespace string) *rbacv1.RoleBinding {
```

File: pkg/subctl/cmd/root.go [hunk 4]
```diff
@@ -112,16 +112,27 @@ func handleNodeLabels(config *rest.Config) error {
...
 			fmt.Printf("* No worker node found to label as the gateway\n")
+			nodes, err := clientset.CoreV1().Nodes().List(context.TODO(), metav1.ListOptions{})
+			if err != nil {
+				return err
+			}
+			if len(nodes.Items) == 0 {
+				return fmt.Errorf("There are no nodes in the cluster")
+			}
+			node = nodes.Items[0].GetName()
+			fmt.Printf("* Labelling %s as gateway\n", node)
```
Comment (skitt) [inline pkg/subctl/cmd/root.go]: This seems unrelated to the broker namespace change, or am I missing something?

File: pkg/subctl/cmd/deploybroker.go [hunk 5]
```diff
@@ -73,6 +74,8 @@ func init() {
 
 	deployBroker.PersistentFlags().BoolVar(&operatorDebug, "operator-debug", false, "enable operator debugging (verbose logging)")
 
+	deployBroker.PersistentFlags().StringVar(&brokerNamespace, "broker-namespace", broker.SubmarinerBrokerNamespace, "namespace for broker")
```
Comment (skitt) [inline pkg/subctl/cmd/deploybroker.go]: I think we should move `broker.SubmarinerBrokerNamespace` here, and make it a tool-specific default. We can pretty much get rid of `pkg/broker/namespace.go` entirely given that `NewBrokerNamespace()` is trivial (it can be merged in `CreateNewBrokerNamespace()`).

**Review by skitt — CHANGES_REQUESTED** [review]
Thanks for submitting this!

I have a few changes I’d like to see, mostly stylistic (repeated parameter types). There seems to be an unrelated change re. node labels.

File: pkg/subctl/cmd/root.go (hunk 4 above)
Comment (donaldh) [reply pkg/subctl/cmd/root.go]: Apologies, this should not be part of the commit. Is there an easy way to remove this from the PR, or should I submit a new PR?

Comment (skitt) [reply pkg/subctl/cmd/root.go]: You can amend your commit and force push your branch; you’ll need to do that anyway because the additional commit generated by GitHub with my suggestions doesn’t carry a DCO... If you’d like help with the git commands feel free to ask.

Comment (donaldh) [reply pkg/subctl/cmd/root.go]: Thx, that worked.

**Review by skitt — APPROVED** [review]

**Review by dfarrell07 — APPROVED** [review]
LGTM, thanks! The linter's complaints:

```
pkg/broker/ensure.go:158:1: paramTypeCombine: func(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (secret *v1.Secret, err error) could be replaced with func(clientset *kubernetes.Clientset, submarinerBrokerSA, namespace string) (secret *v1.Secret, err error) (gocritic)
func WaitForClientToken(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (secret *v1.Secret, err error) {
^
pkg/broker/ensure.go:209:1: paramTypeCombine: func(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (brokerSA *v1.ServiceAccount, err error) could be replaced with func(clientset *kubernetes.Clientset, submarinerBrokerSA, namespace string) (brokerSA *v1.ServiceAccount, err error) (gocritic)
func CreateNewBrokerSA(clientset *kubernetes.Clientset, submarinerBrokerSA string, namespace string) (brokerSA *v1.ServiceAccount, err error) {
^
pkg/subctl/operator/brokercr/ensure.go:41: File is not `gofmt`-ed with `-s` (gofmt)
```

File: pkg/subctl/cmd/deploybroker.go [hunk 6]
```diff
@@ -38,6 +38,10 @@ import (
 	"github.com/submariner-io/submariner-operator/pkg/subctl/operator/submarinerop"
 )
 
+const (
+	DefaultBrokerNamespace = "submariner-k8s-broker"
```
Comment (tpantelis) [inline pkg/subctl/cmd/deploybroker.go]: This doesn't need to be exported so let's not leak it.

```suggestion
	defaultBrokerNamespace = "submariner-k8s-broker"
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Review by tpantelis — APPROVED** [review]

**Review by sridhargaddam — APPROVED** [review]

**Comment by nyechiel** [comment]
@donaldh @dfarrell07 can you please ensure the new flag is documented?

**Comment by donaldh** [comment]
Hi Nir,

Yes, I will do that. It will likely be January before I get to it. On holiday now. 

Thanks,
Donald

**Comment by nyechiel** [comment]
Thanks @donaldh. I reported this issue so that we don't forget about this. Feel free to assign to yourself (I think it can only be assigned to you after you add a comment there): https://github.com/submariner-io/submariner-website/issues/630

Enjoy the break!

File: pkg/subctl/cmd/deploybroker.go [hunk 7]
```diff
@@ -73,6 +78,8 @@ func init() {
 
 	deployBroker.PersistentFlags().BoolVar(&operatorDebug, "operator-debug", false, "enable operator debugging (verbose logging)")
 
+	deployBroker.PersistentFlags().StringVar(&brokerNamespace, "broker-namespace", defaultBrokerNamespace, "namespace for broker")
```
Comment (mkolesnik) [inline pkg/subctl/cmd/deploybroker.go:81]: I know this was merged a while ago, but isn't it simpler UX and more consistent to just name this "--namespace"?

I'm talking about the flag name not the variable name.

Comment (donaldh) [reply pkg/subctl/cmd/deploybroker.go:81]: The name was chosen to make it clear that it sets the broker namespace, as opposed to the operator namespace, or more generally to avoid suggesting the command's effect would only apply to a single `--namespace`

Comment (mkolesnik) [reply pkg/subctl/cmd/deploybroker.go:81]: Yes that might make sense, but considering we're talking about `subctl deploy-broker` command I think having `broker` in the flag as well is repetitive.

Also, aren't we essentially affecting a single namespace here? So I'm not sure it's incorrect in that regard.
